src/flint/flint_base/flint_base.pyx

```diff
@@ -30,7 +30,127 @@ cdef class flint_elem:
 
 
 cdef class flint_scalar(flint_elem):
-    pass
+    # =================================================
+    # These are the functions a new class should define
+    # assumes that addition and multiplication are
+    # commutative
+    # =================================================
+    def is_zero(self):
+        return False
+
+    def _any_as_self(self):
+        return NotImplemented
+
+    def _neg_(self):
+        return NotImplemented
+
+    def _add_(self, other):
+        return NotImplemented
+
+    def _sub_(self, other, swap=False):
+        return NotImplemented
+
+    def _mul_(self, other):
+        return NotImplemented
+
+    def _div_(self, other, swap=False):
+        return NotImplemented
+
+    def _floordiv_(self, other, swap=False):
+        return NotImplemented
+
+    def _invert_(self):
+        return NotImplemented
+
+    # =================================================
+    # Generic arithmetic using the above functions
+    # =================================================
+
+    def __pos__(self):
+        return self
+
+    def __neg__(self):
+        return self._neg_()
+
+    def __add__(self, other):
+        other = self._any_as_self(other)
+        if other is NotImplemented:
+            return NotImplemented
+        return self._add_(other)
+
+    def __radd__(self, other):
+        other = self._any_as_self(other)
+        if other is NotImplemented:
+            return NotImplemented
+        return self._add_(other)
+
+    def __sub__(self, other):
+        other = self._any_as_self(other)
+        if other is NotImplemented:
+            return NotImplemented
+        return self._sub_(other)
+
+    def __rsub__(self, other):
+        other = self._any_as_self(other)
+        if other is NotImplemented:
+            return NotImplemented
+        return self._sub_(other, swap=True)
```

Comment:
Rather than having `swap=True` I would just have two methods `_sub_` and `_rsub_`.

Reply:
The idea is that `__sub__` handles all conditionality so that `_sub_` and `_rsub_` do not need to check anything. Also keyword arguments can add overhead that I would avoid at this level.

Reply:
OK I've done this for the base class and added `rsub`, `rdiv`.